# Project Milestone Two: Modeling and Feature Engineering

### Due: Midnight on April 13 (with 2-hour grace period) and worth 25 points

### Overview

This milestone builds on your work from Milestone 1. You will:

1. Evaluate baseline models using default settings.
2. Engineer new features and re-evaluate models.
3. Use feature selection techniques to find promising subsets.
4. Select the top 3 models and fine-tune them for optimal performance.

You must do all work in this notebook and upload to your team leader's account in Gradescope. There is no
Individual Assessment for this Milestone. 


In [1]:
# ===================================
# Useful Imports: Add more as needed
# ===================================

# Standard Libraries
import os
import time
import math
import io
import zipfile
import requests
from urllib.parse import urlparse
from itertools import chain, combinations

# Data Science Libraries
import numpy as np
import pandas as pd
import seaborn as sns

# Visualization
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import matplotlib.ticker as mticker  # Optional: Format y-axis labels as dollars
import seaborn as sns

# Scikit-learn (Machine Learning)
from sklearn.model_selection import (
    train_test_split, 
    cross_val_score, 
    GridSearchCV, 
    RandomizedSearchCV, 
    RepeatedKFold
)
from sklearn.preprocessing import StandardScaler, OrdinalEncoder
from sklearn.impute import SimpleImputer
from sklearn.metrics import mean_squared_error
from sklearn.feature_selection import SequentialFeatureSelector, f_regression, SelectKBest
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor, GradientBoostingRegressor

# Progress Tracking

from tqdm import tqdm

# =============================
# Global Variables
# =============================
random_state = 42

# =============================
# Utility Functions
# =============================

# Format y-axis labels as dollars with commas (optional)
def dollar_format(x, pos):
    return f'${x:,.0f}'

# Convert seconds to HH:MM:SS format
def format_hms(seconds):
    return time.strftime("%H:%M:%S", time.gmtime(seconds))



## Prelude: Load your Preprocessed Dataset from Milestone 1

In Milestone 1, you handled missing values, encoded categorical features, and explored your data. Before you begin this milestone, you’ll need to load that cleaned dataset and prepare it for modeling.

Here’s what to do:

1. Return to your Milestone 1 notebook and rerun your code through Part 3, where your dataset was fully cleaned (assume it’s called `df_cleaned`).

2. **Save** the cleaned dataset to a file by running:

>   df_cleaned.to_csv("zillow_cleaned.csv", index=False)

3. Switch to this notebook and **load** the saved data:

>   df = pd.read_csv("zillow_cleaned.csv")

4. Create a **train/test split** using `train_test_split`.  
   
6. **Standardize** the features (but not the target!) using **only the training data.** This ensures consistency across models without introducing data leakage from the test set:

>   scaler = StandardScaler()   
>   X_train_scaled = scaler.fit_transform(X_train)    
  
**Notes:** 

- You will not use the testing set during this milestone — it’s reserved for final evaluation later.
- You will have to redo the scaling step when you introduce new features (which have to be scaled as well).


In [2]:
df = pd.read_csv("zillow_cleaned.csv")

In [3]:
df.isna().sum().sum()

np.int64(0)

In [4]:
# Add as many code cells as you need
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X = df.drop('taxvaluedollarcnt', axis=1)  
y = df['taxvaluedollarcnt']  


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)




### Part 1: Baseline Modeling [3 pts]

Apply the following regression models to the scaled training dataset using **default parameters**:

- Linear Regression
- Ridge Regression
- Lasso Regression
- Decision Tree Regression
- Bagging
- Random Forest
- Gradient Boosting Trees

For each model:
- Use **repeated cross-validation** (e.g., 5 folds, 5 repeats).
- Report the **mean and standard deviation of CV RMSE Score** across all folds in a table. 


In [14]:
# Add as many code cells as you need

#Linear Regression

import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, RepeatedKFold


model = LinearRegression()

cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

linear_reg_results = {
    'Model': 'Linear Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}




In [15]:
# Ridge

from sklearn.linear_model import Ridge

model = Ridge()

cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

ridge_reg_results = {
    'Model': 'Ridge Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [16]:
# Lasso
from sklearn.linear_model import Lasso

model = Lasso()

cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

lasso_reg_results = {
    'Model': 'Lasso Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

/home/codespace/.local/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 8.104e+15, tolerance: 2.701e+12
  model = cd_fast.enet_coordinate_descent(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 6.833e+15, tolerance: 2.313e+12
  model = cd_fast.enet_coordinate_descent(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duali

In [17]:
# Decision Tree 

from sklearn.tree import DecisionTreeRegressor

model = DecisionTreeRegressor()

cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

decision_tree_results = {
    'Model': 'Decision Tree Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [18]:
# Bagging 

from sklearn.ensemble import BaggingRegressor

model = BaggingRegressor()
cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

bagging_results = {
    'Model': 'Bagging Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [19]:
# Random Forest 

from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(n_estimators=10, max_depth=10, n_jobs=-1, random_state=42)

cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

random_forest_results = {
    'Model': 'Random Forest Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [21]:
#Gradient Boosting

from sklearn.ensemble import GradientBoostingRegressor

model = GradientBoostingRegressor(random_state=42)

cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X, y, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

gradient_boosting_results = {
    'Model': 'Gradient Boosting Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [22]:
results = [
    linear_reg_results,
    ridge_reg_results,
    lasso_reg_results,
    decision_tree_results,
    bagging_results,
    random_forest_results,
    gradient_boosting_results
]

results_df = pd.DataFrame(results)
print(results_df)


                         Model      Mean RMSE      Std RMSE
0            Linear Regression  510577.122282  55441.756243
1             Ridge Regression  509605.296441  55120.462340
2             Lasso Regression  510561.302332  55437.863581
3      Decision Tree Regressor  656212.210158  64116.453111
4            Bagging Regressor  497609.669099  50081.377901
5      Random Forest Regressor  490369.367314  50403.779666
6  Gradient Boosting Regressor  485880.754959  52187.774282


### Part 1: Discussion [2 pts]

In a paragraph or well-organized set of bullet points, briefly compare and discuss:

  - Which models perform best overall?
  - Which are most stable (lowest std)?
  - Any signs of overfitting or underfitting?

To determine which model performed best, we looked at Mean RMSE of each model; the lower this value, the better the model performed. Therefore, according to our results, Gradient Boosting Regressor (Mean RMSE: 485880.754959) was the best model in part 1. However, this score is still a very large number and is concerning in terms of our model's accuracy in predicting the target value. When looking at which model is the most stable (or has the lowest standard deviation), we see the Bagging Regressor the best option. This model has a STD of 50081.377901 which indictates it fits the data the most consistently (compared to the other models) and there are not major outliers effecting the performance of the model. Based on the results, we would say the Decision Tree Regression model is overfitting the data, which typically is the a trend in this model. The model resulted in a large Mean RMSE value and a large STD which signifies it may not be capturing the complexity of the training data and is not consistent in where the errors occur. This model will mostly likely not generalize well to unseen data.

### Part 2: Feature Engineering [3 pts]

Consider **at least three new features** based on your Milestone 1, Part 5. Examples include:
- Polynomial terms
- Log or interaction terms
- Groupings or transformations of categorical features

Add these features to `X_train` and then:
- Scale using `StandardScaler` 
- Re-run all models listed above (using default settings again).
- Report updated RMSE scores (mean and std) across repeated CV in a table. 

**Note:**  Recall that this will require creating a new version of the dataset, so effectively you may be running "polynomial regression" using `LinearRegression`. 

In [23]:
# Add as many code cells as you need

from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2, include_bias=False)
X_poly = poly.fit_transform(X_train)

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_poly_scaled = scaler.fit_transform(X_poly)



In [25]:
# Linear Regression and Polynomial Features

model = LinearRegression()

cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)

cv_results = cross_val_score(model, X_poly_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)  
std_rmse = np.std(cv_results)

# Store results in dictionary
linear_regression_results = {
    'Model': 'Linear Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [26]:
# Ridge Regression model and Polynomial Features

model = Ridge()
cv_results = cross_val_score(model, X_poly_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

ridge_regression_results = {
    'Model': 'Ridge Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [27]:
# Lasso Regression model and Polynomial Features

model = Lasso()

cv_results = cross_val_score(model, X_poly_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

lasso_regression_results = {
    'Model': 'Lasso Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}



/home/codespace/.local/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 5.056e+15, tolerance: 1.882e+12
  model = cd_fast.enet_coordinate_descent(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 4.629e+15, tolerance: 1.775e+12
  model = cd_fast.enet_coordinate_descent(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/linear_model/_coordinate_descent.py:697: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duali

In [28]:
# Decision Tree Regressor model and Polynomial Features

model = DecisionTreeRegressor()
cv_results = cross_val_score(model, X_poly_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

# Store results in dictionary
decision_tree_results_poly = {
    'Model': 'Decision Tree Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [29]:
# Bagging Regressor model and Polynomial Features

model = BaggingRegressor()
cv_results = cross_val_score(model, X_poly_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
bagging_regressor_results_poly = {
    'Model': 'Bagging Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [ ]:
# Random Forest Regressor model and Polynomial Features

import time
total_iterations = 10000000
start_time = time.time()
for i in range(total_iterations):
    if i % (total_iterations // 10) == 0:  
        elapsed_time = time.time() - start_time
        estimated_remaining_time = (elapsed_time / (i + 1)) * (total_iterations - (i + 1))
        print(f"Progress: {i / total_iterations * 100:.2f}% | Estimated remaining time: {estimated_remaining_time:.2f} seconds")

model = RandomForestRegressor()
cv_results = cross_val_score(model, X_poly_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

random_forest_results_poly = {
    'Model': 'Random Forest Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

KeyboardInterrupt: 

In [ ]:
# Gradient Boosting Regressor model and Polynomial Features

model = GradientBoostingRegressor()

cv_results = cross_val_score(model, X_poly_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)


mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

gradient_boosting_results_poly = {
    'Model': 'Gradient Boosting Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


Progress: 0.00% | Estimated remaining time: 860.69 seconds
Progress: 10.00% | Estimated remaining time: 1.03 seconds
Progress: 20.00% | Estimated remaining time: 0.88 seconds
Progress: 30.00% | Estimated remaining time: 0.77 seconds
Progress: 40.00% | Estimated remaining time: 0.65 seconds
Progress: 50.00% | Estimated remaining time: 0.64 seconds
Progress: 60.00% | Estimated remaining time: 0.58 seconds
Progress: 70.00% | Estimated remaining time: 0.48 seconds
Progress: 80.00% | Estimated remaining time: 0.31 seconds
Progress: 90.00% | Estimated remaining time: 0.15 seconds


In [ ]:
# Results of all models with Polynomial Features 

all_results = [
    linear_regression_results,
    ridge_regression_results,
    lasso_regression_results,
    decision_tree_results_poly,
    bagging_regressor_results_poly,
    random_forest_results_poly,
    gradient_boosting_results_poly
]

results_df = pd.DataFrame(all_results)

print(results_df)


In [ ]:
X_log = np.log1p(X_train)  
scaler = StandardScaler()
X_log_scaled = scaler.fit_transform(X_log)

In [ ]:
# Linear Regression + Log

model = LinearRegression()
cv = RepeatedKFold(n_splits=5, n_repeats=5, random_state=42)
cv_results = cross_val_score(model, X_log_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
linear_regression_log_results = {
    'Model': 'Linear Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Ridge + Log 
model = Ridge()
cv_results = cross_val_score(model, X_log_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
ridge_regression_log_results = {
    'Model': 'Ridge Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [ ]:
# Lasso + Log
from sklearn.linear_model import Lasso
model = Lasso()
cv_results = cross_val_score(model, X_log_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

lasso_regression_log_results = {
    'Model': 'Lasso Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Decision Tree + Log

model = DecisionTreeRegressor()
cv_results = cross_val_score(model, X_log_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
decision_tree_log_results = {
    'Model': 'Decision Tree Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Bagging + Log

model = BaggingRegressor()
cv_results = cross_val_score(model, X_log_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
bagging_log_results = {
    'Model': 'Bagging Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Random Forest + Log

model = RandomForestRegressor()
cv_results = cross_val_score(model, X_log_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
random_forest_log_results = {
    'Model': 'Random Forest Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Gradient Boosting + Log
model = GradientBoostingRegressor()
cv_results = cross_val_score(model, X_log_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
gradient_boosting_log_results = {
    'Model': 'Gradient Boosting Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Results of all models with Log-Transformation
all_log_results = [
    linear_regression_log_results,
    ridge_regression_log_results,
    lasso_regression_log_results,
    decision_tree_log_results,
    bagging_log_results,
    random_forest_log_results,
    gradient_boosting_log_results
]

log_results_df = pd.DataFrame(all_log_results)

print(log_results_df)


In [ ]:
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

encoder = OneHotEncoder(sparse=False, drop='first') 
X_encoded = encoder.fit_transform(X_train)
scaler = StandardScaler()
X_encoded_scaled = scaler.fit_transform(X_encoded)

In [ ]:
# Linear Regression + One-Hot Encoding 
model = LinearRegression()
cv_results = cross_val_score(model, X_encoded_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
linear_regression_encoded_results = {
    'Model': 'Linear Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [ ]:
# Ridge + One-Hot Encoding 

model = Ridge()
cv_results = cross_val_score(model, X_encoded_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)
ridge_regression_encoded_results = {
    'Model': 'Ridge Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Lasso + One-Hot Encoding 

model = Lasso()
cv_results = cross_val_score(model, X_encoded_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

lasso_regression_encoded_results = {
    'Model': 'Lasso Regression',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}

In [ ]:
# Decision Tree + One-Hot Encoding 

model = DecisionTreeRegressor()

cv_results = cross_val_score(model, X_encoded_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

decision_tree_encoded_results = {
    'Model': 'Decision Tree Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}



In [ ]:
# Bagging + One-Hot Encoding 

model = BaggingRegressor()

cv_results = cross_val_score(model, X_encoded_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)
mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

# Store results in dictionary
bagging_encoded_results = {
    'Model': 'Bagging Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}




In [ ]:
# Random Forest + One-Hot Encoding 

model = RandomForestRegressor()

cv_results = cross_val_score(model, X_encoded_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

random_forest_encoded_results = {
    'Model': 'Random Forest Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}



In [ ]:
# Gradient Boosting + One-Hot Encoding 
model = GradientBoostingRegressor()

cv_results = cross_val_score(model, X_encoded_scaled, y_train, scoring='neg_root_mean_squared_error', cv=cv, n_jobs=-1)

mean_rmse = -np.mean(cv_results)
std_rmse = np.std(cv_results)

gradient_boosting_encoded_results = {
    'Model': 'Gradient Boosting Regressor',
    'Mean RMSE': mean_rmse,
    'Std RMSE': std_rmse
}


In [ ]:
# Results of all models with One-Hot Encoding 
all_encoded_results = [
    linear_regression_encoded_results,
    ridge_regression_encoded_results,
    lasso_regression_encoded_results,
    decision_tree_encoded_results,
    bagging_encoded_results,
    random_forest_encoded_results,
    gradient_boosting_encoded_results
]

encoded_results_df = pd.DataFrame(all_encoded_results)
print(encoded_results_df)


### Part 2: Discussion [2 pts]

Reflect on the impact of your new features:

- Did any models show notable improvement in performance?

- Which new features seemed to help — and in which models?

- Do you have any hypotheses about why a particular feature helped (or didn’t)?

- Were there any unexpected results?



- Did any models show notable improvement in performance?

- Which new features seemed to help — and in which models?

- Do you have any hypotheses about why a particular feature helped (or didn’t)?

- Were there any unexpected results?
    

### Part 3: Feature Selection [3 pts]

Using the full set of features (original + engineered):
- Apply **feature selection** methods to investigate whether you can improve performance.
  - You may use forward selection, backward selection, or feature importance from tree-based models.
- For each model, identify the **best-performing subset of features**.
- Re-run each model using only those features.
- Report updated RMSE scores (mean and std) across repeated CV in a table.


### Part 3: Discussion [2 pts]

Analyze the effect of feature selection on your models:

- Did performance improve for any models after reducing the number of features?

- Which features were consistently retained across models?

- Were any of your newly engineered features selected as important?

- How did feature selection differ between linear and tree-based models?

> Your text here

### Part 4: Fine-Tuning Your Top 3 Models [6 pts]

In this final phase of Milestone 2, you’ll select and refine your **three most promising models and their corresponding data pipelines** based on everything you've done so far.

1. Choose the top 3 models based on performance and interpretability from earlier parts.
2. For each model:
   - Perform hyperparameter tuning using `sweep_parameters`, `GridSearchCV`, `RandomizedSearchCV`, or other techniques from previous homeworks. 
   - Experiment with different versions of your feature engineering and preprocessing — treat these as additional tunable components.
3. Report the mean and standard deviation of CV RMSE score for each model in a summary table.



In [ ]:
# Add as many code cells as you need

### Part 4: Discussion [4 pts]

Reflect on your tuning process and final results:

- What was your tuning strategy for each model? Why did you choose those hyperparameters?
- Did you find that certain types of preprocessing or feature engineering worked better with specific models?
- Provide a ranking of your three models and explain your reasoning — not just based on RMSE, but also interpretability, training time, or generalizability.
- Conclude by considering whether this workflow has produced the results you expected. Typically, you would repeat steps 2 - 4 and also reconsider the choices you made in Milestone 1 when cleaning the dataset, until reaching the point of diminishing returns; do you think that would that have helped here?

> Your text here